framework/src/positions/ElementGroupCentroidPositions.C

```diff
@@ -0,0 +1,351 @@
+      return volumes_2d[indices[0]];
+    else if (_extra_id_names.size() == 3)
+      return volumes_3d[indices[0]][indices[1]];
+    else
+      return volumes_4d[indices[0]][indices[1]][indices[2]];
+  };
+
+  std::vector<std::map<unsigned int, unsigned int>> positions_indexing(
+      _extra_id_group_indices.size());
+
+  // Make index maps to go from the extra element id to the positions array index
+  for (MooseIndex(_extra_id_group_indices) i = 0; i < _extra_id_group_indices.size(); i++)
+  {
+    auto & indices = _extra_id_group_indices[i];
+    auto j = 0;
+    for (auto extra_id : indices)
+      positions_indexing[i][extra_id] = j++;
+  }
+
+  for (const auto & elem : _mesh.getMesh().active_element_ptr_range())
+  {
+    // Pre-compute the centroid, this is expensive but may be used for multiple element ids
+    const auto centroid = elem->true_centroid();
+    const auto volume = elem->volume();
+
+    // Keeps track of indices in multi-D arrays
+    std::vector<unsigned int> previous_indices(4);
+
+    for (MooseIndex(_extra_id_names) i = 0; i < _extra_id_names.size(); i++)
+    {
+      auto iter =
+          positions_indexing[i].find(id(*elem, _extra_id_indices[i], _blocks_in_use && (i == 0)));
+      if (iter == positions_indexing[i].end())
+        break;
+      else if (_extra_id_names.size() == i + 1)
+      {
+        getNestedPositions(previous_indices)[iter->second] += volume * centroid;
+        getNestedVolumes(previous_indices)[iter->second] += volume;
+        break;
+      }
+      else
+        previous_indices[i] = iter->second;
+    }
+  }
+
+  // Report the zero volumes
+  unsigned int num_zeros = 0;
+  if (_extra_id_names.size() == 1)
+    for (const auto & vol : volumes)
+      if (MooseUtils::absoluteFuzzyEqual(vol, 0))
+        num_zeros++;
+  if (_extra_id_names.size() == 2)
+    for (const auto & vol_vec : volumes_2d)
+      for (const auto & vol : vol_vec)
+        if (MooseUtils::absoluteFuzzyEqual(vol, 0))
+          num_zeros++;
+  if (_extra_id_names.size() == 3)
+    for (const auto & vol_vec_vec : volumes_3d)
+      for (const auto & vol_vec : vol_vec_vec)
+        for (const auto & vol : vol_vec)
+          if (MooseUtils::absoluteFuzzyEqual(vol, 0))
+            num_zeros++;
+  if (_extra_id_names.size() == 4)
+    for (const auto & vol_vec_vec_vec : volumes_4d)
+      for (const auto & vol_vec_vec : vol_vec_vec_vec)
+        for (const auto & vol_vec : vol_vec_vec)
+          for (const auto & vol : vol_vec)
+            if (MooseUtils::absoluteFuzzyEqual(vol, 0))
+              num_zeros++;
+  if (num_zeros)
+    mooseWarning(std::to_string(num_zeros) +
+                 " zero volume bins detected during group centroid position calculation. "
+                 "The corresponding positions will be removed from consideration.");
+
+  // Renormalize by the total bin volumes
+  if (_extra_id_names.size() == 1)
+    for (MooseIndex(_positions) i = 0; i < _positions.size(); i++)
```

Comment:
I would replace all uses of `MooseIndex` with `index_range`

Reply:
will do

Reply:
Does index_range work with a delete nested?
I get crashes

Reply:
Oh yea erasing invalidates the end iterator so can't use the `index_range` in that case